### PR TITLE
Annotate System.Net.Sockets for nullable reference types

### PR DESCRIPTION
**Pull request by eerhardt**
Contributes to #2339

Depends on #32518.

/cc @dotnet/ncl

### CODE REVIEW

File: src/libraries/System.Net.Sockets/ref/System.Net.Sockets.cs [hunk 1]
```diff
@@ -258,27 +259,27 @@ public partial class Socket : System.IDisposable
...
+        public System.IAsyncResult BeginSend(byte[] buffer, int offset, int size, System.Net.Sockets.SocketFlags socketFlags, System.AsyncCallback? callback, object? state) { throw null; }
+        public System.IAsyncResult? BeginSend(byte[] buffer, int offset, int size, System.Net.Sockets.SocketFlags socketFlags, out System.Net.Sockets.SocketError errorCode, System.AsyncCallback? callback, object? state) { throw null; }
+        public System.IAsyncResult BeginSend(System.Collections.Generic.IList<System.ArraySegment<byte>> buffers, System.Net.Sockets.SocketFlags socketFlags, System.AsyncCallback? callback, object? state) { throw null; }
+        public System.IAsyncResult? BeginSend(System.Collections.Generic.IList<System.ArraySegment<byte>> buffers, System.Net.Sockets.SocketFlags socketFlags, out System.Net.Sockets.SocketError errorCode, System.AsyncCallback? callback, object? state) { throw null; }
+        public System.IAsyncResult BeginSendFile(string? fileName, System.AsyncCallback? callback, object? state) { throw null; }
```
Comment (eerhardt) [inline src/libraries/System.Net.Sockets/ref/System.Net.Sockets.cs]: I wasn't sure how to handle `SendFile(string fileName)`and  `BeginSendFile(string fileName, ...)`. Technically the code allows for the `fileName` to be null - but if you aren't sending a file, and these overloads have no `pre|postBuffer`, then what exactly are you sending here?

**Comment by danmosemsft** [comment]
Please help add exactly one area label if the bot does not. It teaches the bot.

**Comment by eerhardt** [comment]
> Please help add exactly one area label if the bot does not. It teaches the bot.

Any thoughts on having the bot post a comment when it doesn't add a label? It could mention that it couldn't add a label, maybe log which areas it _thinks_ are the closest, and ask a human to manually add a label.

**Comment by eerhardt** [comment]
@eiriktsarpalis @buyaa-n - any feedback?

File: src/libraries/Common/src/Interop/Windows/WinSock/Interop.TransmitFile.cs [hunk 2]
```diff
@@ -1,7 +1,7 @@
 // Licensed to the .NET Foundation under one or more agreements.
 // The .NET Foundation licenses this file to you under the MIT license.
 // See the LICENSE file in the project root for more information.
-
+#nullable enable
```
Comment (buyaa-n) [inline src/libraries/Common/src/Interop/Windows/WinSock/Interop.TransmitFile.cs]: We normally empty line in between license header and `#nullable enable`

File: src/libraries/Common/src/Interop/Windows/WinSock/Interop.setsockopt.cs [hunk 3]
```diff
@@ -1,7 +1,7 @@
 // Licensed to the .NET Foundation under one or more agreements.
 // The .NET Foundation licenses this file to you under the MIT license.
 // See the LICENSE file in the project root for more information.
-
+#nullable enable
```
Comment (buyaa-n) [inline src/libraries/Common/src/Interop/Windows/WinSock/Interop.setsockopt.cs]: same here and everywhere else

File: src/libraries/Common/src/System/Net/ContextAwareResult.cs [hunk 4]
```diff
@@ -158,7 +159,7 @@ internal ExecutionContext ContextCopy
                     {
                         NetEventSource.Fail(this, "Must lock (StartPostingAsyncOp()) { ... FinishPostingAsyncOp(); } when calling ContextCopy (unless it's only called after FinishPostingAsyncOp).");
```
Comment (buyaa-n) [inline src/libraries/Common/src/System/Net/ContextAwareResult.cs]: I think we can add to this method declaration [DoesNotReturn], then we can avoid `!` below.

File: src/libraries/Common/src/System/Net/LazyAsyncResult.cs [hunk 5]
```diff
@@ -253,7 +251,7 @@ internal bool InternalPeekCompleted
         }
 
         // Internal property for setting the IO result.
-        internal object Result
+        internal object? Result
```
Comment (buyaa-n) [inline src/libraries/Common/src/System/Net/LazyAsyncResult.cs]:
``` csharp
    if (value == DBNull.Value)
    {
        NetEventSource.Fail(this, "Result can't be set to DBNull - it's a special internal value.");
    }
```
Makes me think its not allow null, no?

File: src/libraries/System.IO.Pipes/src/System/IO/Pipes/NamedPipeClientStream.Unix.cs [hunk 6]
```diff
@@ -28,7 +28,7 @@ private bool TryConnect(int timeout, CancellationToken cancellationToken)
             SafePipeHandle? clientHandle = null;
             try
             {
-                socket.Connect(new UnixDomainSocketEndPoint(_normalizedPipePath));
+                socket.Connect(new UnixDomainSocketEndPoint(_normalizedPipePath!));
```
Comment (buyaa-n) [inline src/libraries/System.IO.Pipes/src/System/IO/Pipes/NamedPipeClientStream.Unix.cs]: Assume it is intended

File: src/libraries/System.Net.Sockets/src/System/Net/Sockets/AcceptOverlappedAsyncResult.Unix.cs [hunk 7]
```diff
@@ -14,9 +14,9 @@ namespace System.Net.Sockets
     // AcceptOverlappedAsyncResult - used to take care of storage for async Socket BeginAccept call.
     internal sealed partial class AcceptOverlappedAsyncResult : BaseOverlappedAsyncResult
     {
-        private Socket _acceptedSocket;
+        private Socket? _acceptedSocket;
 
-        internal Socket AcceptSocket
+        internal Socket? AcceptSocket
```
Comment (buyaa-n) [inline src/libraries/System.Net.Sockets/src/System/Net/Sockets/AcceptOverlappedAsyncResult.Unix.cs]: NIT: not sure why its need to be nullable

File: src/libraries/System.Net.Sockets/src/System/Net/Sockets/AcceptOverlappedAsyncResult.Windows.cs [hunk 8]
```diff
@@ -40,9 +40,10 @@ internal override object PostCompletion(int numBytes)
                     safeHandle.DangerousAddRef(ref refAdded);
                     IntPtr handle = safeHandle.DangerousGetHandle();
 
+                    byte[] buffer = _buffer!;
```
Comment (buyaa-n) [inline src/libraries/System.Net.Sockets/src/System/Net/Sockets/AcceptOverlappedAsyncResult.Windows.cs]: Isn't asserting is better?

File: src/libraries/System.Net.Sockets/src/System/Net/Sockets/AcceptOverlappedAsyncResult.Windows.cs [hunk 9]
```diff
@@ -112,13 +113,14 @@ private void LogBuffer(long size)
             // may fire erroneously.
             Debug.Assert(NetEventSource.IsEnabled);
 
+            byte[] buffer = _buffer!;
```
Comment (buyaa-n) [inline src/libraries/System.Net.Sockets/src/System/Net/Sockets/AcceptOverlappedAsyncResult.Windows.cs]: I would assert

File: src/libraries/Common/src/Interop/Windows/WinSock/Interop.TransmitFile.cs [hunk 10]
```diff
@@ -14,7 +14,7 @@ internal static partial class Mswsock
         [DllImport(Interop.Libraries.Mswsock, SetLastError = true)]
         internal static extern unsafe bool TransmitFile(
             SafeHandle socket,
-            SafeHandle fileHandle,
+            SafeHandle? fileHandle,
```
Comment (stephentoub) [inline src/libraries/Common/src/Interop/Windows/WinSock/Interop.TransmitFile.cs]: In what situation do we pass null here?

File: src/libraries/Common/src/System/Net/LazyAsyncResult.cs (hunk 5 above)
Comment (stephentoub) [reply src/libraries/Common/src/System/Net/LazyAsyncResult.cs]: DBNull is different from null.

Comment (stephentoub) [reply src/libraries/Common/src/System/Net/LazyAsyncResult.cs]: (That said, I wouldn't spend too much thought on this file... I'm hopeful we can delete it outright in the not too distant future.)

File: src/libraries/System.Net.Sockets/ref/System.Net.Sockets.cs [hunk 11]
```diff
@@ -258,27 +259,27 @@ public partial class Socket : System.IDisposable
...
+        public System.IAsyncResult BeginConnect(System.Net.EndPoint remoteEP, System.AsyncCallback? callback, object? state) { throw null; }
+        public System.IAsyncResult BeginConnect(System.Net.IPAddress address, int port, System.AsyncCallback? requestCallback, object? state) { throw null; }
+        public System.IAsyncResult BeginConnect(System.Net.IPAddress[] addresses, int port, System.AsyncCallback? requestCallback, object? state) { throw null; }
+        public System.IAsyncResult BeginConnect(string host, int port, System.AsyncCallback? requestCallback, object? state) { throw null; }
+        public System.IAsyncResult BeginDisconnect(bool reuseSocket, System.AsyncCallback? callback, object? state) { throw null; }
+        public System.IAsyncResult BeginReceive(byte[] buffer, int offset, int size, System.Net.Sockets.SocketFlags socketFlags, System.AsyncCallback? callback, object? state) { throw null; }
+        public System.IAsyncResult? BeginReceive(byte[] buffer, int offset, int size, System.Net.Sockets.SocketFlags socketFlags, out System.Net.Sockets.SocketError errorCode, System.AsyncCallback? callback, object? state) { throw null; }
```
Comment (stephentoub) [inline src/libraries/System.Net.Sockets/ref/System.Net.Sockets.cs]: I could not have told you that some of these BeginXx methods return null.  That's quite unfortunate.  I'm glad this pattern is deprecated :)

File: src/libraries/System.Net.Sockets/ref/System.Net.Sockets.cs (hunk 1 above)
Comment (stephentoub) [reply src/libraries/System.Net.Sockets/ref/System.Net.Sockets.cs]: I agree.  I think this is a case where, even though we're not actually validating, the intent is for the string to be non-null, and we can say it's non-nullable.  If we end up finding a use case, we can always relax it in the future.

But I think that only applies to this overload.  For the others, we have no way to express that at least one of the three inputs should be not null but the others could be, so I think we need to leave them all as nullable.

File: src/libraries/System.Net.Sockets/ref/System.Net.Sockets.cs [hunk 12]
```diff
@@ -348,8 +349,8 @@ public partial class Socket : System.IDisposable
         public int Send(System.ReadOnlySpan<byte> buffer, System.Net.Sockets.SocketFlags socketFlags) { throw null; }
         public int Send(System.ReadOnlySpan<byte> buffer, System.Net.Sockets.SocketFlags socketFlags, out System.Net.Sockets.SocketError errorCode) { throw null; }
         public bool SendAsync(System.Net.Sockets.SocketAsyncEventArgs e) { throw null; }
-        public void SendFile(string fileName) { }
-        public void SendFile(string fileName, byte[] preBuffer, byte[] postBuffer, System.Net.Sockets.TransmitFileOptions flags) { }
+        public void SendFile(string? fileName) { }
```
Comment (stephentoub) [inline src/libraries/System.Net.Sockets/ref/System.Net.Sockets.cs]: As with your previous comment, this should probably be non-nullable.

File: src/libraries/System.Net.Sockets/ref/System.Net.Sockets.cs [hunk 13]
```diff
@@ -644,14 +646,14 @@ public partial class UdpClient : System.IDisposable
         public void JoinMulticastGroup(System.Net.IPAddress multicastAddr) { }
         public void JoinMulticastGroup(System.Net.IPAddress multicastAddr, int timeToLive) { }
         public void JoinMulticastGroup(System.Net.IPAddress multicastAddr, System.Net.IPAddress localAddress) { }
-        public byte[] Receive(ref System.Net.IPEndPoint remoteEP) { throw null; }
+        public byte[] Receive(ref System.Net.IPEndPoint? remoteEP) { throw null; }
```
Comment (stephentoub) [inline src/libraries/System.Net.Sockets/ref/System.Net.Sockets.cs]: Should `remoteEP` here have a `[NotNull]` attribute on it? i.e. we allow it to be null when passed in, but it'll never be null on the way out?

File: src/libraries/System.Net.Sockets/src/System/Net/Sockets/Socket.cs [hunk 14]
```diff
@@ -3862,6 +3865,11 @@ public bool DisconnectAsync(SocketAsyncEventArgs e)
             // Throw if socket disposed
             ThrowIfDisposed();
 
+            if (e == null)
+            {
+                throw new ArgumentNullException(nameof(e));
+            }
```
Comment (stephentoub) [inline src/libraries/System.Net.Sockets/src/System/Net/Sockets/Socket.cs]: You may have already done this and I just haven't come to it yet, but can you make sure there's a test for this?

**Review by stephentoub — APPROVED** [review]
Skimmed through it and left a few comments, but generally LGTM.  Welcome to the wonderful world of nullable reference types ;)

File: src/libraries/System.Net.Sockets/src/System/Net/Sockets/MulticastOption.cs [hunk 15]
```diff
@@ -25,7 +25,7 @@ public MulticastOption(IPAddress group, IPAddress mcint)
                 throw new ArgumentNullException(nameof(mcint));
             }
 
-            Group = group;
+            _group = group;
```
Comment (eiriktsarpalis) [inline src/libraries/System.Net.Sockets/src/System/Net/Sockets/MulticastOption.cs]: Why this change?

Comment (eiriktsarpalis) [inline src/libraries/System.Net.Sockets/src/System/Net/Sockets/MulticastOption.cs]: Why this change?

File: src/libraries/System.Net.Sockets/src/System/Net/Sockets/SocketAsyncEventArgs.Windows.cs [hunk 16]
```diff
@@ -21,27 +21,27 @@ private enum SingleBufferHandleState : byte { None, InProcess, Set }
...
+        private Internals.SocketAddress? _pinnedSocketAddress;
 
         // SendPacketsElements property variables.
-        private FileStream[] _sendPacketsFileStreams;
+        private FileStream[]? _sendPacketsFileStreams;
 
         // Overlapped object related variables.
-        private PreAllocatedOverlapped _preAllocatedOverlapped;
-        private readonly StrongBox<SocketAsyncEventArgs> _strongThisRef = new StrongBox<SocketAsyncEventArgs>(); // state for _preAllocatedOverlapped; .Value set to this while operations in flight
+        private PreAllocatedOverlapped _preAllocatedOverlapped = null!; // initialized by helper called from ctor
```
Comment (eiriktsarpalis) [inline src/libraries/System.Net.Sockets/src/System/Net/Sockets/SocketAsyncEventArgs.Windows.cs]: I know this is the standard way of forcing non-nullability of fields, but I think it might become an issue if/when new constructors are being added or modified: the compiler will provide no warnings if the field is not being initialized. Do you think it might make sense to move those forced initializations into the individual constructors that need it? It would also make it easier to audit in the future. cc @buyaa-n @stephentoub

**Review by eiriktsarpalis — APPROVED** [review]

File: src/libraries/System.Net.Sockets/src/System/Net/Sockets/SocketAsyncEventArgs.Windows.cs (hunk 16 above)
Comment (stephentoub) [reply src/libraries/System.Net.Sockets/src/System/Net/Sockets/SocketAsyncEventArgs.Windows.cs]: The right way to handle this will be with the `[MemberNotNull]` attribute that's coming.  I'd prefer to wait for that rather than changing the scheme we employ now.  I also don't think the duplication of all that initialization would be worthwhile.

File: src/libraries/System.Net.Sockets/src/System/Net/Sockets/MulticastOption.cs (hunk 15 above)
Comment (buyaa-n) [reply src/libraries/System.Net.Sockets/src/System/Net/Sockets/MulticastOption.cs]: Because compiler complains that the field `_group ` not initialized

**Review by buyaa-n — APPROVED** [review]
Thanks!

File: src/libraries/Common/src/Interop/Windows/WinSock/Interop.TransmitFile.cs (hunk 2 above)
Comment (eerhardt) [reply src/libraries/Common/src/Interop/Windows/WinSock/Interop.TransmitFile.cs]: Fixed all of them.

File: src/libraries/Common/src/System/Net/ContextAwareResult.cs (hunk 4 above)
Comment (eerhardt) [reply src/libraries/Common/src/System/Net/ContextAwareResult.cs]: I initially had that, but then backed off since this method does return, so I thought it was more correct to not decorate it with `[DoesNotReturn]`.

I can go either way here. I'll add the attribute.

File: src/libraries/System.IO.Pipes/src/System/IO/Pipes/NamedPipeClientStream.Unix.cs (hunk 6 above)
Comment (eerhardt) [reply src/libraries/System.IO.Pipes/src/System/IO/Pipes/NamedPipeClientStream.Unix.cs]: Yes, this change is needed. I logged https://github.com/dotnet/runtime/issues/32760 to track this scenario.

File: src/libraries/System.Net.Sockets/src/System/Net/Sockets/AcceptOverlappedAsyncResult.Unix.cs (hunk 7 above)
Comment (eerhardt) [reply src/libraries/System.Net.Sockets/src/System/Net/Sockets/AcceptOverlappedAsyncResult.Unix.cs]: The one place that sets this property sets it to a `Socket?`. And on Unix it is always `null`.

https://github.com/dotnet/runtime/blob/03fc3ab4f63e6bbb99f300769b786e72007d98ff/src/libraries/System.Net.Sockets/src/System/Net/Sockets/Socket.cs#L3504

File: src/libraries/Common/src/System/Net/ContextAwareResult.cs (hunk 4 above)
Comment (stephentoub) [reply src/libraries/Common/src/System/Net/ContextAwareResult.cs]: Doesn't it always call Debug.Fail? From a nullability analysis perspective then, it would seem to be just as much a DoesNotReturn as Debug.Fail,no?

File: src/libraries/System.Net.Sockets/src/System/Net/Sockets/AcceptOverlappedAsyncResult.Windows.cs (hunk 8 above)
Comment (eerhardt) [reply src/libraries/System.Net.Sockets/src/System/Net/Sockets/AcceptOverlappedAsyncResult.Windows.cs]: Agreed. Fixed.

File: src/libraries/System.Net.Sockets/src/System/Net/Sockets/AcceptOverlappedAsyncResult.Windows.cs (hunk 9 above)
Comment (eerhardt) [reply src/libraries/System.Net.Sockets/src/System/Net/Sockets/AcceptOverlappedAsyncResult.Windows.cs]: Fixed.

File: src/libraries/Common/src/Interop/Windows/WinSock/Interop.TransmitFile.cs (hunk 10 above)
Comment (eerhardt) [reply src/libraries/Common/src/Interop/Windows/WinSock/Interop.TransmitFile.cs]: In the case where `BeginSendFile(string? fileName, byte[]? preBuffer, byte[]? postBuffer` is called with a null or empty `fileName`.

File: src/libraries/System.Net.Sockets/ref/System.Net.Sockets.cs (hunk 11 above)
Comment (eerhardt) [reply src/libraries/System.Net.Sockets/ref/System.Net.Sockets.cs]: I was surprised when I saw this as well. A win for the nullable annotation work :) to make this more expressive.

File: src/libraries/Common/src/System/Net/ContextAwareResult.cs (hunk 4 above)
Comment (eerhardt) [reply src/libraries/Common/src/System/Net/ContextAwareResult.cs]: Good point. I didn't realize `Debug.Fail` had `[DoesNotReturn]`. Fixed.

File: src/libraries/System.Net.Sockets/ref/System.Net.Sockets.cs (hunk 13 above)
Comment (eerhardt) [reply src/libraries/System.Net.Sockets/ref/System.Net.Sockets.cs]: Good call. Added.

File: src/libraries/System.Net.Sockets/src/System/Net/Sockets/Socket.cs (hunk 14 above)
Comment (eerhardt) [reply src/libraries/System.Net.Sockets/src/System/Net/Sockets/Socket.cs]: No, I forgot. Thanks for the reminder.

**Comment by eerhardt** [comment]
Thanks for all the feedback. I have addressed it, and if I don't receive anymore I'll merge this when it is green.

**Comment by eerhardt** [comment]
The 2 OSX failures are #32882.
